packages/extension-base/src/services/keyring-service/context/handlers/Migration.ts

```diff
@@ -0,0 +1,249 @@
+      } else { // derived account
+        if (notMigrateDeriveBefore) { // todo: can be optimized later
+          await new Promise((resolve) => setTimeout(resolve, 2000));
+          notMigrateDeriveBefore = false;
+        }
+
+        const suri = unifiedAccount.suri || '';
+
+        this.parentService.context.derivationAccountProxyCreate({ name, suri, proxyId: unifiedAccount.parentId || '' }, true);
+        unifiedAccountIds.push(proxyId);
+      }
+    }
+
+    keyring.lockAll(false);
+    this.parentService.updateKeyringState();
+
+    return unifiedAccountIds;
+  }
+
+  public getUACanBeMigrated (accountProxies: AccountProxy[]): AccountProxy[] {
+    return accountProxies.filter((account) => this.state.isUnifiedAccount(account.id) && account.isNeedMigrateUnifiedAccount);
+  }
+
+  public getEligibleSoloAccounts (accountProxies: AccountProxy[]): AccountProxy[] {
+    return accountProxies.filter((account) => !this.state.isUnifiedAccount(account.id) && account.isNeedMigrateUnifiedAccount);
+  }
+
+  public groupSoloAccountByPair (password: string, accountProxies: AccountProxy[]) {
+    const parentService = this.parentService;
+
+    return accountProxies.reduce(function (rs: Record<string, AccountProxy[]>, item) {
+      const oldProxyId = item.id;
+      const mnemonic = parentService.context.exportAccountProxyMnemonic({
+        password,
+        proxyId: oldProxyId
+      }).result;
+      const upcomingProxyId = createAccountProxyId(mnemonic);
+
+      if (!rs[upcomingProxyId]) {
+        rs[upcomingProxyId] = [];
+      }
+
+      rs[upcomingProxyId].push(item);
+
+      return rs;
+    }, {});
+  }
+
+  public accountProxiesToEligibleSoloAccountMap (accountProxyMap: Record<string, AccountProxy[]>): Record<string, SoloAccountToBeMigrated[]> {
+    const eligibleSoloAccountMap: Record<string, SoloAccountToBeMigrated[]> = {};
+
+    Object.entries(accountProxyMap).forEach(([upcomingProxyId, accounts]) => {
+      eligibleSoloAccountMap[upcomingProxyId] = accounts.map((account) => {
+        return {
+          upcomingProxyId,
+          proxyId: account.accounts[0].proxyId,
+          address: account.accounts[0].address,
+          name: account.name,
+          chainType: account.chainTypes[0]
+        } as SoloAccountToBeMigrated;
+      });
+    });
+
+    return eligibleSoloAccountMap;
+  }
+
+  public sortUAByParent (accountProxies: AccountProxy[]): AccountProxy[] {
+    const undefinedToStr = (str: string | undefined) => str ?? '';
+
+    return accountProxies.sort((a, b) => undefinedToStr(a.parentId) < undefinedToStr(b.parentId) ? -1 : undefinedToStr(a.parentId) > undefinedToStr(b.parentId) ? 1 : 0);
+  }
+
+  public migrateSoloToUnifiedAccount (request: RequestMigrateSoloAccount): ResponseMigrateSoloAccount {
+    const { accountName, sessionId, soloAccounts } = request;
+    const password = this.sessionIdToPassword[sessionId].password;
+
+    keyring.unlockKeyring(password);
+    this.parentService.updateKeyringState();
```

Comment:
why do we need this?

Reply:
We lock keyring after done migrate unified accounts -> unified account. Need to unlock again to migrate solo account. The flow is like this:
1. Unlock keyring -> Migrate all unified accounts -> Lock keyring -> 
2.1. Unlock keyring -> Migrate solo account -> Lock keyring ->
2.2. Unlock keyring -> Migrate solo account 2 -> Lock keyring ->
2.n.Unlock keyring -> Migrate solo account n -> Lock keyring